File: benchmark/comparison/tasks/29/solution.ts
type Result<T> = { tag: "Ok"; value: T } | { tag: "Err"; message: string };

function parse_int(s: string): Result<number> {
  const n = Number(s);
  if (isNaN(n) || s.trim() === "") return { tag: "Err", message: `not a number: ${s}` };
  if (!Number.isInteger(n)) return { tag: "Err", message: `not a number: ${s}` };
  return { tag: "Ok", value: n };
}

function compute(a: Result<number>, b: Result<number>): Result<number> {
  if (a.tag === "Err") return a;
  if (b.tag === "Err") return b;
  return { tag: "Ok", value: a.value + b.value };
}

function describe_result(r: Result<number>): string {
  if (r.tag === "Ok") return `Success: ${r.value}`;
  return `Error: ${r.message}`;
}

function formatResult(r: Result<number>): string {
  if (r.tag === "Ok") return `Ok(${r.value})`;
  return `Err(${r.message})`;
}

console.log(formatResult(parse_int("42")));
console.log(formatResult(parse_int("abc")));

console.log(describe_result(compute({ tag: "Ok", value: 3 }, { tag: "Ok", value: 4 })));
console.log(describe_result(compute({ tag: "Err", message: "bad" }, { tag: "Ok", value: 4 })));
console.log(describe_result(compute({ tag: "Ok", value: 3 }, { tag: "Err", message: "bad" })));

console.log(describe_result(compute(parse_int("42"), parse_int("10"))));
console.log(describe_result(compute(parse_int("42"), parse_int("abc"))));

File: benchmark/comparison/tasks/29/solution.py
def parse_int(s):
    try:
        return ("Ok", int(s))
    except ValueError:
        return ("Err", f"not a number: {s}")

def compute_add(r1, r2):
    if r1[0] == "Err":
        return r1
    if r2[0] == "Err":
        return r2
    return ("Ok", r1[1] + r2[1])

def describe_result(result):
    if result[0] == "Ok":
        return f"Success: {result[1]}"
    return f"Error: {result[1]}"

def fmt(result):
    if result[0] == "Ok":
        return f"Ok({result[1]})"
    return f"Err({result[1]})"

print(fmt(parse_int("42")))
print(fmt(parse_int("abc")))
print(describe_result(("Ok", 7)))
print(describe_result(("Err", "bad")))
print(describe_result(compute_add(("Err", "bad"), ("Ok", 1))))
print(describe_result(compute_add(parse_int("42"), parse_int("10"))))
print(describe_result(compute_add(parse_int("42"), parse_int("abc"))))
